### ISSUE TITLE
todo

### ISSUE THREAD

**Issue by lucidrains** [lucidrains/titans-pytorch#2]
- [ ] chunking
- [ ] test neural memory with local attention transformer on enwik8, need to replicate the effects of extra MLP layers + momentum
- [ ] special forward that closures past states + time steps (which needs to be tracked due to chunking)
- [ ] take care of past states for associative scans
- [ ] tensor typing
- [ ] explorations
  - [ ] attention in place of the MLP, processing sets of key / values
  - [ ] memorizing multiple key / values per token
  - [ ] factorizing weight matrices in MLP

**Comment by errorsandwarnings** [comment]
@lucidrains The paper looks really good. How are the preliminary results looking like? I can help test it if you need.

**Comment by lucidrains** [comment]
@errorsandwarnings hey, think it should be ready around Sunday

don't get your hopes up though, some issues are already arising in my mind. but overall it is a well written paper that generalizes TTT and recent linear attention architectures. just need to see it before believing

**Comment by lucidrains** [comment]
@errorsandwarnings think it should be ready around Sunday for testing 🤞

**Comment by errorsandwarnings** [comment]
Is there any experiment in your mind we can use to check it against transformer, the results in the paper show this is almost around the board better than transformers. I know this kind of claim is too large but if it is true probably the best results I have seen. The perplexity drop on language modelling is quite high in my opinion. I can support running some comparisons once the code is ready to bring some real numbers.

**Comment by lucidrains** [comment]
my plan is to first make sure it can bring improvements to a severe local attention transformer. then I need to see that adding more layers to the MLP actually does anything that stacking linear attention layers does not. if it passes all those tests, then look into momentum, weight decay etc

**Comment by lucidrains** [comment]
@errorsandwarnings yes I agree the results look good, paper well written, but you never know

**Comment by errorsandwarnings** [comment]
Sounds reasonable.

**Comment by errorsandwarnings** [comment]
> @errorsandwarnings yes I agree the results look good, paper well written, but you never know

Yeah, the never know part :) I really want a cheaper better transformer.

**Comment by errorsandwarnings** [comment]
@lucidrains  How is the implementation coming up? Any observations so far?

**Comment by errorsandwarnings** [comment]
@lucidrains  Take rest, well deserved after such an effort.

**Comment by lucidrains** [comment]
<img width="450" alt="Screenshot 2025-01-13 at 11 43 56 AM" src="https://github.com/user-attachments/assets/3c002d1a-11cc-439e-8e8c-bf080a1ba7d9" />

seeing something, but will require more days to verify

**Comment by errorsandwarnings** [comment]
@lucidrains  How much data? How long did you run?

**Comment by lucidrains** [comment]
it is just character level enwik8, and only for this morning

just want to see it do something and not nothing

**Comment by lucidrains** [comment]
network is kind of sensitive to the adaptive learning rate as well as having an output normalization

the performance also seems to vary depending on the chunk size

also, just to put it out there, i didn't see a straightforward improvement when i cranked the number of MLP layers from 2 to 4

**Comment by errorsandwarnings** [comment]
Yeah, should we run this through authors once completed?

**Comment by errorsandwarnings** [comment]
> network is kind of sensitive to the adaptive learning rate as well as having an output normalization
> 
> the performance also seems to vary depending on the chunk size
> 
> also, just to put it out there, i didn't see a straightforward improvement when i cranked the number of MLP layers from 2 to 4. that was a bit disappointing, but i could be doing something wrong
> 
> will keep playing around with it / polish it up by week's end

Can it not happen if the underlying data/signal is too simple to learn.

**Comment by lucidrains** [comment]
> > network is kind of sensitive to the adaptive learning rate as well as having an output normalization
> > the performance also seems to vary depending on the chunk size
> > also, just to put it out there, i didn't see a straightforward improvement when i cranked the number of MLP layers from 2 to 4. that was a bit disappointing, but i could be doing something wrong
> > will keep playing around with it / polish it up by week's end
> 
> Can it not happen if the underlying data/signal is too simple to learn.

yes it could be, as I wasn't able to run the experiments for that long, and the networks tend to learn the more distant patterns further into training

**Comment by errorsandwarnings** [comment]
@lucidrains  Any further news here?

**Comment by lucidrains** [comment]
@errorsandwarnings yup, spent another morning with experiments and I think I'm still seeing it. Will be wiring up the full MAC attention pattern config tomorrow morning, with the long term and persistent memories

however, still not seeing the big improvements when trying to go up to a 4 layer MLP, at least for the first 3k steps

**Comment by errorsandwarnings** [comment]
@lucidrains Ok, Would it be worth comparing on another dataset ?

**Comment by Hyacinth-YX** [comment]
Thanks for sharing. I'm really curious to see if it works as expected. Looking forward to hearing about the results!

**Comment by sentialx** [comment]
@lucidrains Did you make any optimizations? I'm trying to train titan but compared to transformer it is very slow (1.5s vs 100ms per step)

**Comment by lucidrains** [comment]
@sentialx hey yes, i noticed that last night, will be removing the `max_grad_norm`

and yea, still a bit unoptimized for the moment

i also did a longer run yesterday and taylor series linear attention outperformed titans given enough time, but it is definitely working as it beat local attention only baseline.

**Comment by lucidrains** [comment]
@sentialx give me some time.. at least until end of the week to sort out what is going on with this paper / if there is anything to it.

**Comment by lucidrains** [comment]
![Screenshot from 2025-01-15 09-50-57](https://github.com/user-attachments/assets/55af610f-8813-40df-8522-1b92b8d958c5)

ok yea, the effects are much clearer on the MAC configuration, where attention is completely localized to segments (here i did 32). i'm not sure why MAC is so well performing in their paper.. usually in local attention, you want overlapping windows so receptive field increases through depth. very weird

**Comment by errorsandwarnings** [comment]
@lucidrains The gap is very prominent here. Looks much better.

**Comment by lucidrains** [comment]
@errorsandwarnings yea it is, wasn't as obvious in my local-attention implementation as there the windows overlap

ok, let me do a quick run with taylor series linear attention, bracing myself to be disappointed

**Comment by errorsandwarnings** [comment]
@lucidrains We should also prepare another baseline + dataset to be sure as well. The plots if they keep going down can be dramatically different on large data.

**Comment by kw2828** [comment]
@lucidrains happy to assist with another baseline + dataset as well. keep up the good fight

**Comment by lucidrains** [comment]
![Screenshot from 2025-01-15 10-19-02](https://github.com/user-attachments/assets/9b026338-ad37-4bae-a7c6-a09c216e9d8f)

aw ok :disappointed:

**Comment by sentialx** [comment]
Mind trying gated deltanet? I found titan to be much weaker than gated delta net

**Comment by errorsandwarnings** [comment]
@sentialx  @lucidrains I think we need to check the "expressive model" theory. Ideally we can run a fairly long LLM comparison - perhaps TinyLLaMA's dataset.

**Comment by lucidrains** [comment]
nevermind I had a bug in the attention code, rerunning

**Comment by lucidrains** [comment]
<img width="1065" alt="Screenshot 2025-01-15 at 12 28 39 PM" src="https://github.com/user-attachments/assets/22fc8aa9-c0a5-4127-a66a-6d3df12654b7" />

ok, bar any implementation issue, i think i'm finally seeing it. very close to going all in on this paper

**Comment by errorsandwarnings** [comment]
May be parallel run a MLP 8 to confirm the trend? And yes this looks better.

**Comment by errorsandwarnings** [comment]
@lucidrains Sounds good, see you tomorrow.

**Comment by sentialx** [comment]
@lucidrains I think what you experienced in MAC with chunking vs SWA is that chunking forces the model (or neural memory) to actually compress information between segments, but that's just a guess

**Comment by lucidrains** [comment]
@sentialx yes, that's what I've been musing with Fares over signal. It is testable!

**Comment by errorsandwarnings** [comment]
@lucidrains  How is the training speed btw of current version ? vs baseline and taylor attention.

**Comment by jobox05** [comment]
Looks really promising; crossing my fingers for long term AI memory this time :p

**Comment by sentialx** [comment]
Not sure why, but for me SWA works better. Also @lucidrains what's the importance of persistent kv keys when we already have longterm memory keys prepended to the sequence?

**Comment by lucidrains** [comment]
@sentialx hey Eryk, want to chat through Signal? better to give you my frank thoughts on the paper there

**Comment by sentialx** [comment]
Sure, my username is sentialx.80

**Comment by errorsandwarnings** [comment]
@lucidrains  How are things looking now?

**Comment by lucidrains** [comment]
@errorsandwarnings pretty good, still seeing it on experiments this morning, hacking away at the sparse attention mask and will get inference done tomorrow

**Comment by kanghuazhao** [comment]
What an absolutely brilliant framework! This is a game-changing breakthrough that finally makes true AI memory possible. I'm deeply impressed by your quick implementation and excited to see where your continued research takes this!

**Comment by henryclw** [comment]
This is so intriguing! I'm going to run the experiments as well.

**Comment by wrench1997** [comment]
May I ask how the experiment went?

**Comment by lucidrains** [comment]
@wrench1997 what i'm looking for is replication of figure 8

<img width="341" alt="Image" src="https://github.com/user-attachments/assets/deabba0a-0268-4609-a397-18b1c9241ec3" />

so far @sentialx has shown me results for 2 layer MLP beating gated delta net, but i haven't seen 3-4 yet. the reason why 2 layer MLP is not impressive is because TTT paper has already shown this..

**Comment by lucidrains** [comment]
really there are only two things new in this paper, momentum and greater MLP depth. momentum can be added to other types of networks (including gated delta net), so it isn't that interesting to me (although it is a nice addition). MLP depth is interesting because it would signal that there is room to explore other higher capacity networks. however, i take it as an implicit negative result that the author says the same thing in his future works section, rather than finding a network that lead to dramatically better perplexity

**Comment by lucidrains** [comment]
@wrench1997 in other words, not conclusive yet.. give us more time, or find a beefy GPU and run some experiments of your own

**Comment by sustcsonglin** [comment]
Hi @lucidrains @sentialx Could I take a look at the results? I am very curious


> [@wrench1997](https://github.com/wrench1997) what i'm looking for is replication of figure 7
> 
> <img alt="Image" width="1021" src="https://private-user-images.githubusercontent.com/108653/405264123-983b76b2-28e4-4fbd-a9c0-3d71d2b90c2f.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mzc0NzQwNjYsIm5iZiI6MTczNzQ3Mzc2NiwicGF0aCI6Ii8xMDg2NTMvNDA1MjY0MTIzLTk4M2I3NmIyLTI4ZTQtNGZiZC1hOWMwLTNkNzFkMmI5MGMyZi5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwMTIxJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDEyMVQxNTM2MDZaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0xOGY2NjZjNWI5NWU2NjdhMjdiZGVkY2JmMzhhOGY1Y2Y5MTJmNzZmMTk3YjE3OTVmNjVmMTVmYjk0YjgwMzg3JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.VXvtT-Whan6gnh_xOz8wWu0jwrj3w1scFp8PMXBf82I">
> so far [@sentialx](https://github.com/sentialx) has shown me results for 2 layer MLP beating gated delta net, but i haven't seen 3-4 yet. the reason why 2 layer MLP is not impressive is because TTT paper has already shown this..

**Comment by lucidrains** [comment]
@sustcsonglin oh hey Songlin! yes, you two should connect and start dialectics

**Comment by sustcsonglin** [comment]
> [@sustcsonglin](https://github.com/sustcsonglin) oh hey Songlin! yes, you two should connect and start dialectics. Eryk shared the results with me over Signal yesterday

Sent you a 'hi' on Signal. How about a group chat there

**Comment by vtsouval** [comment]
@lucidrains I can also run some experiments and share the results. Is there any specific setup you want me to test?

**Comment by lucidrains** [comment]
@vtsouval hey yea sure! want to hit me up on Signal? so just as a correction @sentialx had a small labeling error and while 2 MLP converges, it did not beat GDN. 3 MLP is still worse

i'm basically looking for any set of hyperparameters under which we can see clearly figure 7 (perplexity drops as depth increases)

**Comment by cemag1** [comment]
@lucidrains how are things looking? tried getting in touch on Signal through the QR code but seems it doesn't work. I have a few ideas I would like to discuss :)

**Comment by scm-aiml** [comment]
@lucidrains was looking through the paper again, for my own edification just going to try an outsiders-code review. Are you still focusing on MAC for right now?
